core/crypto/aes-gcm.c

```diff
@@ -0,0 +1,306 @@
+	if (res)
+		return res;
+
+	if (nonce_len == (96 / 8)) {
+		memcpy(ctx->ctr, nonce, nonce_len);
+		aes_gcm_inc_ctr(ctx);
+	} else {
+		ghash_update_pad_zero(ctx, nonce, nonce_len);
+		ghash_update_lengths(ctx, 0, nonce_len);
+
+		memcpy(ctx->ctr, ctx->hash_state, sizeof(ctx->ctr));
+		memset(ctx->hash_state, 0, sizeof(ctx->hash_state));
+	}
+
+	aes_gcm_encrypt_block(ctx, ctx->ctr, ctx->buf_tag);
+	aes_gcm_inc_ctr(ctx);
+	if (mode == TEE_MODE_ENCRYPT) {
+		/*
+		 * Encryption uses the pre-encrypted xor-buffer to encrypt
+		 * while decryption encrypts the xor-buffer when needed
+		 * instead.
+		 *
+		 * The reason for this is that the combined encryption and
+		 * ghash implementation does both operations intertwined.
+		 * In the decrypt case the xor-buffer is needed at the end
+		 * of processing each block, while the encryption case
+		 * needs xor-buffer before processing each block.
+		 *
+		 * In a pure software implementation we wouldn't have any
+		 * use for this kind of optimization, but since this
+		 * AES-GCM implementation is aimed at being combined with
+		 * accelerated routines it's more convenient to always have
+		 * this optimization activated.
+		 */
+		aes_gcm_encrypt_block(ctx, ctx->ctr, ctx->buf_cryp);
+		aes_gcm_inc_ctr(ctx);
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result aes_gcm_update_aad(struct aes_gcm_ctx *ctx, const void *data,
+			      size_t len)
+{
+	const uint8_t *d = data;
+	size_t l = len;
+	const uint8_t *head = NULL;
+	size_t n;
+
+	if (ctx->payload_bytes)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	ctx->aad_bytes += len;
+
+	while (l) {
+		if (ctx->buf_pos || !ptr_is_block_aligned(d) ||
+		    l < TEE_AES_BLOCK_SIZE) {
+			n = MIN(TEE_AES_BLOCK_SIZE - ctx->buf_pos, l);
+			memcpy(ctx->buf_hash + ctx->buf_pos, d, n);
+			ctx->buf_pos += n;
+
+			if (ctx->buf_pos != TEE_AES_BLOCK_SIZE)
+				return TEE_SUCCESS;
+
+			ctx->buf_pos = 0;
+			head = ctx->buf_hash;
+			d += n;
+			l -= n;
+		}
+
+		if (ptr_is_block_aligned(d))
+			n = l / TEE_AES_BLOCK_SIZE;
+		else
+			n = 0;
+
+		aes_gcm_ghash_update(ctx, head, d, n);
+		l -= n * TEE_AES_BLOCK_SIZE;
+		d += n * TEE_AES_BLOCK_SIZE;
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result aes_gcm_update_payload(struct aes_gcm_ctx *ctx, TEE_OperationMode m,
+				  const void *src, size_t len, void *dst)
+{
+	size_t n;
+	const uint8_t *s = src;
+	uint8_t *d = dst;
+	size_t l = len;
+
+	if (!ctx->payload_bytes && ctx->buf_pos) {
+		/* AAD part done, finish up the last bits. */
+		memset(ctx->buf_hash + ctx->buf_pos, 0,
+		       TEE_AES_BLOCK_SIZE - ctx->buf_pos);
+		aes_gcm_ghash_update(ctx, ctx->buf_hash, NULL, 0);
+		ctx->buf_pos = 0;
+	}
+
+	ctx->payload_bytes += len;
+
+	while (l) {
+		if (ctx->buf_pos || !ptr_is_block_aligned(s) ||
+		    !ptr_is_block_aligned(d) || l < TEE_AES_BLOCK_SIZE) {
+			n = MIN(TEE_AES_BLOCK_SIZE - ctx->buf_pos, l);
+
+			if (!ctx->buf_pos && m == TEE_MODE_DECRYPT) {
+				aes_gcm_encrypt_block(ctx, ctx->ctr,
+						      ctx->buf_cryp);
```

Comment:
`aes_gcm_encrypt_block()` or `aes_gcm_decrypt_block()` ?

Reply:
Always encrypt, this is the counter in AES Counter mode.